src/v/iceberg/rest_client/catalog_client.cc

```diff
@@ -0,0 +1,187 @@
+/*
+ * Copyright 2024 Redpanda Data, Inc.
+ *
+ * Licensed as a Redpanda Enterprise file under the Redpanda Community
+ * License (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * https://github.com/redpanda-data/redpanda/blob/master/licenses/rcl.md
+ */
+
+#include "iceberg/rest_client/catalog_client.h"
+
+#include "bytes/iobuf_parser.h"
+#include "http/request_builder.h"
+#include "http/utils.h"
+#include "iceberg/rest_client/parsers.h"
+
+#include <seastar/core/sleep.hh>
+#include <seastar/coroutine/as_future.hh>
+
+#include <absl/strings/str_join.h>
+#include <absl/strings/strip.h>
+
+namespace {
+
+template<typename T>
+T trim_slashes(std::optional<T> input, typename T::type default_value) {
+    if (!input.has_value()) {
+        return T{default_value};
+    }
+    return T{absl::StripSuffix(absl::StripPrefix(input.value()(), "/"), "/")};
+}
+
+} // namespace
+
+namespace iceberg::rest_client {
+
+expected<json::Document> parse_json(iobuf&& raw_response) {
+    iobuf_parser p{std::move(raw_response)};
+    auto raw_json = p.read_string(p.bytes_left());
+
+    json::Document doc;
+    doc.Parse(raw_json);
+
+    if (doc.HasParseError()) {
+        return tl::unexpected(json_parse_error{
+          .context = "parse_json",
+          .error = GetParseError_En(doc.GetParseError())});
+    }
+
+    return doc;
+}
+
+catalog_client::catalog_client(
+  http::abstract_client& client,
+  ss::sstring endpoint,
+  credentials credentials,
+  std::optional<base_path> base_path,
+  std::optional<prefix_path> prefix,
+  std::optional<api_version> api_version,
+  std::optional<oauth_token> token,
+  std::unique_ptr<retry_policy> retry_policy)
+  : _client{client}
+  , _endpoint{std::move(endpoint)}
+  , _credentials{std::move(credentials)}
+  , _path_components{std::move(base_path), std::move(prefix), std::move(api_version)}
+  , _oauth_token{std::move(token)}
+  , _retry_policy{
+      retry_policy ? std::move(retry_policy)
+                   : std::make_unique<default_retry_policy>()} {}
+
+ss::future<expected<oauth_token>>
+catalog_client::acquire_token(retry_chain_node& rtc) {
+    const auto token_request
+      = http::request_builder{}
+          .method(boost::beast::http::verb::post)
+          .path(
+            fmt::format("{}oauth/tokens", _path_components.token_api_path()))
+          .header("content-type", "application/x-www-form-urlencoded");
+    auto payload = http::form_encode_data({
+      {"grant_type", "client_credentials"},
+      {"client_id", _credentials.client_id},
+      {"client_secret", _credentials.client_secret},
+      {"scope", "PRINCIPAL_ROLE:ALL"},
```

Comment:
Good point, all examples for the rest catalog use `PRINCIPAL_ROLE:ALL`, this is not documented clearly in the spec, but I assume we might want to scope this to a specific principal role instead of trying to get `ALL`.  

This also leads into the question of how we want to set up the principal and role, should it be done by redpanda or do we assume it is already done and expect a credential id and secret, I assumed the latter, and that the credential id and secret will be part of the environment or config.

We might want to parameterize this scope but we would have to know the principal role value for this. Maybe that should also be config/env variable.